src/codegen/transformations.lisp

```diff
@@ -246,3 +201,169 @@
+                    (type util:symbol-list bound-variables)
+                    (values node-dynamic-extent &optional))
+           (let ((new-args (list ':bound-variables
+                                 (cons (node-dynamic-extent-name node) bound-variables))))
+             (copy-node node
+                        :node (traverse (node-dynamic-extent-node node) funs new-args)
+                        :body (traverse (node-dynamic-extent-body node) funs new-args)))))
+   (cons ':traverse-bind
+         (lambda (node funs &key bound-variables)
+           (declare (type node-bind node)
+                    (type list funs)
+                    (type util:symbol-list bound-variables)
+                    (values node-bind &optional))
+           (let ((new-args (list ':bound-variables
+                                 (cons (node-bind-name node) bound-variables))))
+             (copy-node node
+                        :expr (traverse (node-bind-expr node) funs new-args)
+                        :body (traverse (node-bind-body node) funs new-args)))))))
+
+(defun traverse-with-binding-list (node funs)
+  (declare (type node node)
+           (type list funs)
+           (values node &optional))
+  (traverse node
+            (append funs binding-list-traversals)
+            (list ':bound-variables nil)))
+
+(defmethod tc:apply-substitution (subs (node node))
+  (declare (type tc:substitution-list subs)
+           (values node &optional))
+  (traverse
+   node
+   (list
+    (cons ':after-each
+          (lambda (node)
+            (declare (type node node)
+                     (values node &optional))
+            (copy-node node
+                       :type (tc:apply-substitution subs (node-type node)))))
+    (cons ':after-direct-application
+          (lambda (node)
+            (declare (type node-direct-application node)
+                     (values node-direct-application &optional))
+            (copy-node node
+                       :rator-type (tc:apply-substitution subs (node-direct-application-rator-type node)))))
+    (cons ':after-match
+          (lambda (node)
+            (declare (type node-match node)
+                     (values node-match &optional))
+            (copy-node node
+                       :branches (mapcar
+                                  (lambda (branch)
+                                    (make-match-branch
+                                     :pattern (tc:apply-substitution subs (match-branch-pattern branch))
+                                     :body (match-branch-body branch)))
+                                  (node-match-branches node)))))
+    (cons ':after-while-let
+          (lambda (node)
+            (declare (type node-while-let node)
+                     (values node-while-let &optional))
+            (copy-node node
+                       :pattern (tc:apply-substitution subs (node-while-let-pattern node))))))))
+
+(defmethod tc:type-variables ((node node))
```

Comment:
This will collect all type variables from `node`s and `pattern`s in the tree of `node`. (Should I include something like this as a docstring in the code, or is it too obvious?)

Reply:
does this need to be a method?

and yes, lean toward writing docs.

Reply:
`tc:type-variables` was already a method for collecting type variables out of `tc:ty`, accessor, node, etc. Of course, we could use a different name for a function which collects type variables out of the AST nodes, in which case it wouldn't need to be a method